feat: List & ListItem detail props

ElliotTheHuman opened the pull request:
- Added detailAccessory and detailHoverDisclosure props to `List` and `ListItem`
- Added truncate prop to `ListItem`
- Changed padding on `ListItem` to support detailAccessory prop (i.e. padding will be on child containers or parent container depending on detailAccessory)
- Added detailAccessory and detailHoverDisclosure-related tests 
- Added typography and textColor styled functions to `Truncate`

- Note: The image snapshots changes _should_ be from changing the label for `ListItem` stories. Was originally "Menu Item" 🤦‍♂️  Outside of that, those old snapshots _should_ be identical to previous snapshots. Thinking it might be best to keep the "Menu Item" label just so we can be absolutely sure the layout / padding changes to `ListItem` didn't change anything visually.

### Requirements

Please check the following items are addressed in your pull request (or are not applicable)

- [ ] Includes test coverage for all changes
- [ ] Documentation updated
- [ ] i18n impacts
- [ ] a11y impacts
- [ ] :rotating_light: Check for image-snapshot changes (run `yarn image-snapshots` locally)

ElliotTheHuman commented on packages/components/src/List/ListItemLayout.tsx:
```diff
@@ -65,18 +68,20 @@ export const ListItemLayoutGrid = styled.div``
+  display: flex;
   font-size: ${({ theme: { fontSizes } }) => fontSizes.small};
   font-weight: ${({ theme: { fontWeights } }) => fontWeights.normal};
+  height: ${({ description, height }) =>
+    description ? height + 16 : height}px;
```
Debating whether this should be min-height or height.

min-height would mean that the item container could expand vertically to properly fit content, but that would throw off the specific measurements we had for each density value.

height would keep density values mapped to specific heights, but putting taller items as content / detail could lead to weirdness.

This choice could affect Explore's Field Picker since we're currently using no vertical padding on `TreeItem`, which allows them to cleanly fit a 24px Icon into a 24px TreeItem.

![image](https://user-images.githubusercontent.com/16812885/105404800-3d58da80-5bdf-11eb-92c9-c56cc3546eda.png)


@jhardy

lukelooker replied:
Hmmm...

ElliotTheHuman replied:
Discussed over Meet, but the plan is to try a css grid implementation that allows detail children to be as big as their parent container.

ElliotTheHuman replied:
Resolved in 66d4330

ElliotTheHuman commented on packages/components/src/List/ListItemLayout.tsx:
```diff
@@ -65,18 +68,20 @@ export const ListItemLayoutGrid = styled.div``
-  button,
-  a,
-  & > div {
+  & > button,
+  & > a {
```
Adding a button or link element in the detail was affected by these selectors previously. Making them more specific should solve that.

ElliotTheHuman commented on packages/components/src/List/ListItemLayout.tsx:
```diff
@@ -132,7 +144,9 @@ export const ListItemLayout = styled(ListItemWrapper)`
 
   &:hover {
-    background: ${({ theme: { colors } }) => colors.ui1};
+    /* stylelint-disable */
```
Hit an odd stylelint issue involving space before the bang operator. Definitely open to suggestions on this one if the disabling isn't ideal.

lukelooker replied:
I wonder if you could do something like `detailAccessory === false && colors.ui1` to not upset stylelint?

ElliotTheHuman replied:
Resolved in 66bbd9c

ElliotTheHuman commented on packages/components/src/List/ListItemDetail.tsx:
```diff
@@ -24,11 +24,15 @@
 import styled from 'styled-components'
 
-export const ListItemDetail = styled.div`
+export const ListItemDetail = styled.div.attrs<PaddingProps>((props) => ({
+  ...props,
+  pl: props.pl || 'xsmall',
+}))<PaddingProps>`
+  ${padding}
```
I needed to be able to have padding on the `ListItemDetail` when detailAccessory === true, but NOT when it was false (i.e. the parent layout container takes care of padding if false).

Giving this padding props seemed like the best solution, but could be wrong.

lukelooker replied:
I'd be tempted to make the solution narrower by having an `accessory` prop that just toggles the padding value as needed instead of making `PaddingProps` available any upstream `ListItemDetail` consumers.

ElliotTheHuman replied:
The one problem is that a `ListItem`'s density determines the size of the conditional "padding-right" on `ListItemDetail`. In which case we need two props (i.e. "density" and "accessory") to replace `PaddingProps` props.

I can do that if that's preferable here!

ElliotTheHuman replied:
Accidentally minimized Luke's previous comment (thought it was a dupe). Added it back here:
```
I'd be tempted to make the solution narrower by having an accessory prop that just toggles the 
padding value as needed instead of making PaddingProps available any upstream ListItemDetail consumers.
```

ElliotTheHuman commented on packages/components/src/Truncate/Truncate.tsx:
```diff
@@ -26,18 +26,31 @@
 import { useTooltip } from '../Tooltip'
 
-export interface TruncateProps extends WidthProps {
+export interface TruncateProps
+  extends TypographyProps,
+    WidthProps,
+    CompatibleHTMLProps<HTMLSpanElement> {
```
Felt like `Truncate` could use TypographyProps. My thought was that since it was used in place of Paragraph depending on X prop, it should have similar props to Paragraph.

lukelooker replied:
Probably fair. It's worth noting that `Truncate` uses a `span` underneath but also has `display: block` so it's an odd fit but I can live with it.

ElliotTheHuman replied:
Resolved in 49909b9

ElliotTheHuman replied:
Resolved in 49909b9

lukelooker commented on packages/components/src/List/List.tsx:
```diff
@@ -52,6 +52,19 @@ export interface ListProps
    */
   density?: DensityRamp
 
+  /**
+   * If true, child ListItems' detail elements will appear outside of the item's grey background on hover
+   * In addition, if true, events will not propagate from the detail container
+   * @default false
+   */
+  detailAccessory?: boolean
+
+  /**
+   * If true, child ListItems' detail elements will only appear on hover
+   * @default false
+   */
+  detailHoverDisclosure?: boolean
```
I'm _tempted_ to say that we shouldn't support these on list and only support them directly on `listItem`

ElliotTheHuman replied:
Resolved in c82b064

lukelooker commented on packages/components/src/List/ListItem.tsx:
```diff
@@ -48,6 +56,17 @@ export interface ListItemProps extends CompatibleHTMLProps<HTMLElement> {
    * Detail element placed right of the item children
    */
   detail?: ReactNode
+  /**
+   * If true, the detail elements will appear outside of the item's grey background on hover
+   * In addition, if true, events will not propagate from the detail container
+   * @default false
+   */
+  detailAccessory?: boolean
+  /**
+   * If true, the detail element will only appear on hover
+   * @default false
+   */
+  detailHoverDisclosure?: boolean
```
Crazy idea... what if `detail` took one of two type signatures:

```tsx
interface DetailOptions {
  /**
   * If true, child ListItems' detail will appear outside of the item's grey background on hover
   * In addition, if true, events will not propagate from the detail container
   * @default false
   */
  accessory?: boolean
  /**
   * If true, child ListItems' detail will only appear on hover
   * @default false
   */
  hoverDisclosure?: boolean
}

interface ListItem {
  ...
  detail: ReactNode | { content: ReactNode, options: DetailOptions }
}
```

ElliotTheHuman replied:
Resolved in eaa1187

lukelooker commented on packages/components/src/List/ListItem.tsx:
```diff
@@ -125,60 +169,93 @@ const ListItemInternal: FC<ListItemProps> = (props) => {
-      >
-        {children}
-      </Paragraph>
+      truncate ? (
+        <Truncate
+          fontSize={itemDimensions.labelFontSize}
+          lineHeight={itemDimensions.labelLineHeight}
+        >
+          {children}
+        </Truncate>
+      ) : (
+        <Paragraph
+          fontSize={itemDimensions.labelFontSize}
+          lineHeight={itemDimensions.labelLineHeight}
+        >
+          {children}
+        </Paragraph>
+      )
```
Use dynamic component reference to simplify this some...

```
     const Component = truncate ? Truncate : Paragraph

     (<Component
        fontSize={itemDimensions.labelFontSize}
        lineHeight={itemDimensions.labelLineHeight}
      >
        {children}
      </Component>
      )
```

ElliotTheHuman replied:
Resolved in f838b1b

lukelooker commented on packages/components/src/List/ListItem.tsx:
```diff
@@ -125,60 +169,93 @@ const ListItemInternal: FC<ListItemProps> = (props) => {
+  const renderedDescription =
+    typeof children === 'string' ? (
+      truncate ? (
+        <Truncate color="text2" fontSize="xsmall">
+          {description}
+        </Truncate>
+      ) : (
+        <Paragraph color="text2" fontSize="xsmall">
+          {description}
+        </Paragraph>
```
Dynamic JSX can be used here too.

ElliotTheHuman replied:
Resolved in f838b1b

lukelooker commented on packages/components/src/List/ListItemLayout.tsx:
```diff
@@ -87,20 +92,28 @@ export const ListItemLayout = styled(ListItemWrapper)`
      */
-    padding: ${({ px, py, theme }) =>
-      `${py === '0.375rem' ? py : theme.space[py]} ${theme.space[px]}`};
+    padding: ${({ detailAccessory, px: propsPx, py: propsPy, theme }) => {
+      const pt = propsPy === '0.375rem' ? propsPy : theme.space[propsPy]
+      const pr = detailAccessory ? '0' : theme.space[propsPx]
+      const pb = pt
+      const pl = theme.space[propsPx]
+
+      return `${pt} ${pr} ${pb} ${pl}`
+    }};
```
Probably worth lifting this out into a separate function given the level of complexity within.

ElliotTheHuman replied:
Resolved in 66d4330

lukelooker reviewed (commented):
When `ListItem` has a `detail` & is `selected` with `detailAccesssory` the hover behavior doesn't look quite as-expected:

![image](https://user-images.githubusercontent.com/34253496/105431724-0ac4d700-5c0b-11eb-9918-22f72a2a9fc3.png)

ElliotTheHuman commented on packages/components/src/List/utils/getItemBackgroundColor.ts:
```diff
@@ -48,11 +48,11 @@ export const treeBackgroundColor = ({
 
-  if (disabled) return
+  if (disabled) renderedColor = 'transparent'
   else if (selected && hovered) renderedColor = stateColors.all
   else if (selected) renderedColor = stateColors.selected
   else if (hovered) renderedColor = stateColors.hovered
-  else return
+  else renderedColor = 'transparent'
```
disabled and else needed to output a valid "background-color" CSS value. Otherwise, the browser's default button CSS takes over in `ListItemLayout`.

ElliotTheHuman commented:
Hey Luke! Got the first round of List & ListItem feedback done. Everything was straightforward except the potential grid implementation (see commit 66d4330).

TLDR is that I couldn't get grid to solve for a lot of the confusing pieces, like the conditional padding on the `ListItemDetail` or having the detail sit within or outside of the internal "button" / "a" element. Apologies ahead of time I'm ignorant to any grid-related properties / magic :( It did help organize some of the JSX structure in the `ListItem` component (but did require the addition of more containers within).

lukelooker commented on packages/components/src/List/ListItem.tsx:
```diff
@@ -62,13 +104,20 @@ export interface ListItemProps extends CompatibleHTMLProps<HTMLElement> {
    * - Use **'button'** for items that trigger in page interactions, like displaying a dialog
    */
   itemRole?: 'link' | 'button'
+  /**
+   * Indicates the ListItem is checked
+   */
+  selected?: boolean
```
`current` and `selected` may need to both exist as `current` has a specific ARIA meaning (current page) vs. `selected` - not exactly sure the right answer here but we should discuss / research what the best practice here needs to be. It may be the case that `selected` and/or `current` needs to either-or perhaps? Anyway, topic to ponder a bit.

ElliotTheHuman replied:
Resolved in 9352c1e

lukelooker commented on packages/components/src/List/types.ts:
```diff
@@ -49,3 +49,26 @@ export const itemDimensionKeys = [
+   */
+  hovered?: boolean
+  /**
+   * If true, the Tree/TreeItem will have a darker background color
```
```suggestion
   * If true, the ListItem will have a darker background color
```

ElliotTheHuman replied:
Resolved in c6c8885

lukelooker commented on packages/components/src/List/types.ts:
```diff
@@ -49,3 +49,26 @@ export const itemDimensionKeys = [
+   */
+  disabled?: boolean
+  /**
+   * If true, the Tree/TreeItem will have a light background color
```
```suggestion
   * If true, the ListItem will have a light background color
```

ElliotTheHuman replied:
Resolved in c6c8885

lukelooker commented on packages/components/src/List/types.ts:
```diff
@@ -49,3 +49,26 @@ export const itemDimensionKeys = [
+   */
+  brand?: boolean
+  /**
+   * If true, the Tree/TreeItem will have a "disabled" presentation.
```
```suggestion
   * If true, the ListItem will have a "disabled" presentation.
```

ElliotTheHuman replied:
Resolved in c6c8885

lukelooker commented on packages/components/src/List/types.ts:
```diff
@@ -49,3 +49,26 @@ export const itemDimensionKeys = [
 ]
+
+export interface ItemBackgroundColorProps {
+  /**
+   * Replace the normal grey selected and selected + hovered color with brand colors
+   * @default false
+   */
+  brand?: boolean
```
It occurs to me that `brand` isn't used anywhere else but `keyColor` is supported by `Link` so it'd be more consistent to use that and create an association the the `theme.colors.key`

```suggestion
  /**
   * Replace the normal uiX color for selected and selected + hovered color with key colors
   * @default false
   */
  keyColor?: boolean
```

ElliotTheHuman replied:
Resolved in 4b6f7ad

lukelooker commented on packages/components/src/List/utils/getItemBackgroundColor.ts:
```diff
@@ -26,15 +26,15 @@
 import { itemSelectedColor, Theme } from '@looker/design-tokens'
 import { css } from 'styled-components'
-import { TreeBackgroundStyleProps } from '../types'
+import { ItemBackgroundColorProps } from '../types'
```
Admittedly, it's annoyingly long but I think for consistency's sake we should title this file and it's interface/functions `ListItemBackgroundColorProps`

ElliotTheHuman replied:
👍  7b0516b

ElliotTheHuman replied:
Resolved in 0a34bfa!

However, what do you think of the name `ListItemStateProps`? I know that set of props functionally exists to determine the background color of the `ListItem`, but they are reflective of the "state" of the `ListItem` (outside of `keyColor`).

lukelooker commented on packages/components/src/List/ListItem.story.tsx:
```diff
@@ -96,6 +126,16 @@ export const Link = () => {
   )
 }
 
+const lorem =
+  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa ...
+export const Truncate = Template.bind({})
+Truncate.args = {
+  ...Basic.args,
+  children: lorem,
+  description: lorem,
+  truncate: true,
+}
+
```
Would be good to add an example of `truncate` + `detail` & `icons` to make sure layout plays nice in a worst-case scenario.

ElliotTheHuman replied:
Resolved in c886cb4!

lukelooker reviewed (commented):
Let's chat tomorrow morning about grid challenges here - overall this is looking really great and I think I have an idea or two about how we can perhaps make managing the "accessory" implementation less painful.

ElliotTheHuman commented on packages/components/src/List/ListItem.tsx:
```diff
@@ -125,60 +163,96 @@ const ListItemInternal: FC<ListItemProps> = (props) => {
+
+  const ContentContainer = !disabled && itemRole === 'link' ? 'a' : 'button'
 
   const renderedChildren =
     typeof children === 'string' ? (
       <Paragraph
         fontSize={itemDimensions.labelFontSize}
         lineHeight={itemDimensions.labelLineHeight}
+        truncate={truncate}
       >
         {children}
       </Paragraph>
     ) : (
       children
     )
 
+  const renderedDescription =
+    typeof description === 'string' ? (
+      <Paragraph
+        color="text2"
+        fontSize={itemDimensions.descriptionFontSize}
+        lineHeight={itemDimensions.descriptionLineHeight}
+        truncate={truncate}
+      >
+        {description}
+      </Paragraph>
+    ) : (
+      description
+    )
```
Ended up going with `Paragraph` and conditionally toggling it's `truncate` prop (didn't realize `Paragraph` had a truncate prop in retrospect 🤦 ).

Considering we check the type of the child / description, it felt like wrapping it in a `Paragraph` was the simplest way to get the right dimensions / truncation behavior.

ElliotTheHuman commented on packages/components/src/List/ListItemLayout.tsx:
```diff
@@ -24,133 +24,46 @@
-      color: ${({ theme: { colors } }) => colors.text1};
-    }
-  }
+export const ListItemLayout = styled(ListItemLayoutInternal)`
+  ${(props) => listItemPadding({ accessory: false, ...props })}
```
I thought having the Layout components handle the padding  made the most readability sense (i.e. packaging the conditional padding with the different JSX layouts).

That being said, I could see an argument for moving this up to the `ListItemWrapper` component so that we're a bit more DRY.

ElliotTheHuman commented:
Summary of the layout changes!

- Added `ListItemWrapper` which is essentially the same as the old `ListItemLayout` (i.e. renders the `li` and handles (most of the) list styling 
- Added `ListItemLayout` and `ListItemLayoutAccessory` which represent the different layout options when accessory is off or on, respectively.
- `ListItem`'s primary purpose to distribute props to either `Layout` or `Wrapper` and also construct the "rendered" children (i.e. icon, children, description, detail)

lukelooker reviewed (changes requested):
Ok - a small pile of what I think is mostly minor nits.

If `codecov` complaints look tricky we can go ahead and defer those for later.

Super excited to enable `Tree`/`TreeItem` density support soon! :)

ElliotTheHuman commented on packages/components/src/List/types.ts:
```diff
@@ -49,3 +53,32 @@ export const itemDimensionKeys = [
   'detailFontSize',
 ]
+
+export interface ListItemBackgroundColorProps {
```
Resolved in 0a279378e!

ElliotTheHuman commented on packages/components/src/List/ListItem.tsx:
```diff
@@ -125,60 +153,105 @@ const ListItemInternal: FC<ListItemProps> = (props) => {
+      icon={renderedIcon}
+      px={itemDimensions.px}
+      py={itemDimensions.py}
+    />
```
Resolved in 0a279378e

ElliotTheHuman commented on packages/components/src/Tree/TreeItem.tsx:
```diff
@@ -53,12 +53,12 @@ import {
 
 export interface TreeItemProps
-  extends TreeBackgroundStyleProps,
+  extends ListItemBackgroundColorProps,
```
Resolved in 0a279378e!

ElliotTheHuman commented on packages/components/src/List/types.ts:
```diff
@@ -49,3 +53,36 @@ export const itemDimensionKeys = [
+  selected?: boolean
+}
+
+export interface ListItemStatefulWithHoveredProps
+  extends ListItemStatefulProps {
+  /**
+   * If true, the ListItem will have a light background color
+   * @default false
+   */
+  hovered?: boolean
+}
```
I don't love this name, but can't come up with anything more concise.

I _could_ do something like `ListItemFullStateful` but figured being explicit was probably more important. This component _should_ only be used by things like `TreeStyle` and the `listItemBackground` util since nothing that devs should touch should have a "hovered" prop.

lukelooker replied:
👍  - Perhaps someday the perfect name will pop into my head but I can live with this as-is for now :)

ElliotTheHuman commented on packages/components/src/Tree/Tree.tsx:
```diff
@@ -94,16 +97,16 @@ const TreeLayout: FC<TreeProps> = ({
         className={className}
         border={hasBorder}
-        brand={hasBrandColoring}
+        keyColor={hasKeyColoring}
```
Resolved in 8861783

ElliotTheHuman commented on packages/components/src/Tree/Tree.tsx:
```diff
@@ -94,16 +97,16 @@ const TreeLayout: FC<TreeProps> = ({
         depth: depth + 1,
         detailAccessory: hasDetailAccessory,
         detailHoverDisclosure: hasDetailHoverDisclosure,
+        keyColor: hasKeyColoring,
```
Resolved in 8861783

ElliotTheHuman commented on packages/components/src/List/types.ts:
```diff
@@ -49,3 +53,32 @@ export const itemDimensionKeys = [
+   * If true, the ListItem will have a light background color
+   * @default false
+   */
+  hovered?: boolean
```
Ended up creating another interface since `hovered` is necessary in >1 spots.

Don't love the name I came up with though. Added a comment covering the new interface below.

ElliotTheHuman commented on packages/components/src/List/ListItem.tsx:
```diff
@@ -125,60 +153,105 @@ const ListItemInternal: FC<ListItemProps> = (props) => {
-    <Component
+  const renderedDescription =
+    typeof description === 'string' ? (
+      <Paragraph
```
Should be resolved in b1eb581

ElliotTheHuman commented on packages/components/src/List/ListItemWrapper.tsx:
```diff
@@ -0,0 +1,151 @@
+
+  ${({ focusVisible, theme: { colors } }) =>
+    focusVisible &&
+    `&:focus-within button:after,
```
Left a note about an attempted test at this behavior. See comment below.

ElliotTheHuman commented on packages/components/src/List/ListItem.test.tsx:
```diff
@@ -59,19 +70,33 @@ describe('ListItem', () => {
+    expect(getByTitle('SVG Title Here')).toBeInTheDocument()
+  })
+
+  // At the moment, JSDom doesn't support the pseudo-selector parameter in getComputedStyle
+  xtest('has a key color border on key press', () => {
+    configure({ computedStyleSupportsPseudoElements: true })
+    const { getByRole } = renderWithTheme(<ListItem>Item</ListItem>)
+    const item = getByRole('listitem')
+    fireEvent.keyUp(item, {
+      key: 'Enter',
+    })
+    expect(window.getComputedStyle(item, ':after')).toHaveAttribute(
+      'border',
+      'solid 2px #9785F2'
+    )
```
Based on the [RTL docs](https://testing-library.com/docs/dom-testing-library/api-configuration/#computedstylesupportspseudoelements), the pseudo-selector parameter in getComputedStyle is not supported by JSDom. Afaik, I don't think there's another way to test the ":after" pseudo-element that creates the purple border when focusVisible === true to make codecov happy.

Could be wrong though (hopefully)! For now, leaving an xtest for future development felt like the right call.

lukelooker replied:
Ah... yes, this is a case where we need to use `jest-image-snapshot` to drive a `puppeteer` instance, move focus and then take a picture. 

Suffice it to say a problem for another day :)

lukelooker reviewed (approved):
Such `List`, much wow!

![List](https://media.giphy.com/media/B7o99rIuystY4/giphy-downsized.gif)
